### PR TITLE
UI: Change global search to use fuzzy search API

### PR DESCRIPTION
**Pull request by backspace**
This updates the UI to use the new [fuzzy search API](https://github.com/hashicorp/nomad/blob/068fd431add9a1079c778ed139248e9a3ce7fca2/website/content/api-docs/search.mdx#fuzzy-searching). It’s a drop-in replacement so the `/` shortcut to jump to search is preserved, and results can be cycled through and chosen via arrow keys and the enter key.

using-fuzzy-search

It doesn’t use everything returned by the API:
* deployments and evaluations: these match by id, doesn’t seem like people would know those or benefit from quick navigation to them
* namespaces: doesn’t seem useful with namespaces as they currently function
* scaling policies
* tasks: the response doesn’t include an allocation id, which means they can’t be navigated to in the UI without an additional query
* CSI volumes: aren’t actually returned by the API

Other design choices that might warrant changes:

1. The search results are ordered by jobs, clients, allocations, task groups, and CSI plugins. Is that a sensible order?
2. When it was client-side and only searched jobs and clients, it would show a heading for both groups even when one had 0 results. I kept that behaviour but maybe with five groups it would be preferable to hide them? image
3. This preserves the existing client-side truncation of results at 10 and also adds a `+` to the full count when the response indicates server-side truncation.
_ image
4. Since there’s no API to check the server configuration and know whether the feature has been disabled, this adds another query in `route:application#beforeModel` that acts as feature detection: if the attempt to query fails (`500`), the global search field is hidden.

Upon having added another query on load, I realised that `beforeModel` was being triggered any time `service:router#transitionTo` was being called, which happens upon navigating to a search result, for instance, because of `refreshModel` being present on the region query parameter. This PR adds a check for `transition.queryParamsOnly` and skips rerunning the onload queries (token permissions check, license check, fuzzy search feature detection).

Implementation notes:

* there are changes to unrelated tests to ignore the on-load feature detection query
* some lifecycle-related guards against `undefined` were required to address failures when navigating to an allocation
* the minimum search length of 2 characters is hard-coded as there’s currently no way to determine [`min_term_length`](https://github.com/hashicorp/nomad/blob/0d058e5110e4e5f9a6092c30651a93c755a78dd1/website/content/docs/configuration/search.mdx#search-parameters) in the UI

Future API improvements/wishlist to record elsewhere:
* a way to know if `min_term_length` has been changed from the default of 2
* similarly, a way to know if fuzzy search is disabled entirely without the hackish need to send a fake query
* allocation ids for tasks, or some way to navigate to tasks

FINALLY, there’s a mysterious problem with the Ember test audit comparison where entire test files aren’t running. I spent 1.5 days trying to address this through CI laparoscopy to no avail. Since it’s not affecting CircleCI nor happening locally, I’m submitting this for review with that known problem that I’ll continue trying to address 😞

### CODE REVIEW

File: ui/app/components/global-search/control.js [hunk 1]
```diff
@@ -57,34 +41,85 @@ export default class GlobalSearchControl extends Component {
...
+    const allAllocationResults = results.Matches.allocs || [];
+    const allTaskGroupResults = results.Matches.groups || [];
+    const allCSIPluginResults = results.Matches.plugins || [];
+
+    const jobResults = allJobResults.slice(0, MAXIMUM_RESULTS).map(({ ID: name, Scope: [ namespace, id ]}) => ({
+      type: 'job',
+      id,
+      namespace,
+      label: name,
+    }));
```
Comment (DingoEatingFuzz) [inline ui/app/components/global-search/control.js:65]: This is a really clean way to handle the API response shape.

File: ui/app/components/lifecycle-chart-row.js [hunk 2]
```diff
@@ -26,6 +26,10 @@ export default class LifecycleChartRow extends Component {
 
   @computed('task.lifecycleName')
   get lifecycleLabel() {
+    if (!this.task) {
+      return '';
+    }
+
```
Comment (DingoEatingFuzz) [inline ui/app/components/lifecycle-chart-row.js:32]: This seems unrelated. How'd you end up here?

File: ui/app/routes/application.js [hunk 3]
```diff
@@ -26,35 +26,49 @@ export default class ApplicationRoute extends Route {
...
 
   async beforeModel(transition) {
-    let exchangeOneTimeToken;
+    let promises;
 
-    if (transition.to.queryParams.ott) {
-      exchangeOneTimeToken = this.get('token').exchangeOneTimeToken(transition.to.queryParams.ott);
+    // service:router#transitionTo can cause this to rerun because of refreshModel on
+    // the region query parameter, this skips rerunning the detection/loading queries.
+    if (transition.queryParamsOnly) {
```
Comment (DingoEatingFuzz) [inline ui/app/routes/application.js:33]: TIL. Nice sleuthing here.

File: ui/mirage/config.js [hunk 4]
```diff
@@ -568,6 +568,74 @@ export default function() {
...
+      },
+      Truncations: {
+        allocs: truncatedAllocs.length < truncatedAllocs.length,
+        groups: truncatedGroups.length < transformedGroups.length,
+        jobs: truncatedJobs.length < transformedJobs.length,
+        nodes: truncatedNodes.length < transformedNodes.length,
+        plugins: truncatedPlugins.length < transformedPlugins.length,
+      },
+    }
+  });
```
Comment (DingoEatingFuzz) [inline ui/mirage/config.js:637]: Woooah, super cool to see this emulated in Mirage.

File: ui/tests/acceptance/client-detail-test.js [hunk 5]
```diff
@@ -17,6 +17,10 @@ let clientToken;
 
 const wasPreemptedFilter = allocation => !!allocation.preemptedByAllocation;
 
+function nonSearchPOSTS() {
+  return server.pretender.handledRequests.reject(request => request.url.includes('fuzzy')).filterBy('method', 'POST');
+}
+
```
Comment (DingoEatingFuzz) [inline ui/tests/acceptance/client-detail-test.js:23]: Seeing stuff like this always makes me want to write some sort of lib that makes building this type of behavior on top of Pretender easier.

**Review by DingoEatingFuzz — APPROVED** [review]
The test audit stuff is a mystery, but the code looks great!

File: ui/app/components/lifecycle-chart-row.js (hunk 2 above)
Comment (backspace) [reply ui/app/components/lifecycle-chart-row.js:32]: I don’t totally understand but I was experiencing test failures when navigating to some search results and had to add a couple of null-handlings, I suspect it has something to do with the way the Mirage models are being constructed. With the time constraint I sadly just added the guards without digging further to grasp the underlying issue 😞

Comment (DingoEatingFuzz) [reply ui/app/components/lifecycle-chart-row.js:32]: That's reasonable!

**Comment by backspace** [comment]
the test audit problem seems to have resolved itself? 🙃🎉
